src/nvim/charset.c

```diff
@@ -370,7 +370,8 @@ size_t transstr_buf(const char *const s, char *const buf, const size_t len)
   while (*p != NUL && buf_p < buf_e) {
     const size_t l = (size_t)utfc_ptr2len((const char_u *)p);
     if (l > 1) {
-      if (buf_p + l >= buf_e) {
+      //ignore lengths that do not fit the buffer
+      if (buf_p + l > buf_e) {
```

Comment:
Similar case in line 387 for non printable character. 

The loop in the else case further down could also be fixed. It iterates over all fields of array `pcc` even if it is not filled up (try unicode 0x9f as statusline character). 
```diff

diff --git a/src/nvim/charset.c b/src/nvim/charset.c
index ab20996df..193d192ac 100644
--- a/src/nvim/charset.c
+++ b/src/nvim/charset.c
@@ -380,10 +380,10 @@ size_t transstr_buf(const char *const s, char *const buf, const size_t len)
         memmove(buf_p, p, l);
         buf_p += l;
       } else {
-        for (size_t i = 0; i < ARRAY_SIZE(pcc); i++) {
+        for (size_t i = 0; i < ARRAY_SIZE(pcc) && pcc[i]; i++) {
           char hexbuf[11];
           const size_t hexlen = transchar_hex(hexbuf, pcc[i]);
-          if (buf_p + hexlen >= buf_e) {
+          if (buf_p + hexlen > buf_e) {
             break;
           }
           memmove(buf_p, hexbuf, hexlen);
````

Reply:
I'm sorry but isnt 0x9f the 'ß' ( im not too sure though )? In any case, i will correct line 387 and add a testcase for some non-printable unicode!

Reply:
You can input unicode with `<CTRL-v>u9f<space>` in insert mode or on the command line.

Reply:
I incorporated your comments! u9f was indeed broken; adding the &pcc[i] check resolved that. 
I added a test; please have a look.